File: libs/sdk/src/credentials.ts
import { chmod, mkdir, readFile, writeFile } from 'node:fs/promises';
import { homedir } from 'node:os';
import { join } from 'node:path';

import type { RegisterResult } from './register.js';

export interface MoltNetConfig {
  identity_id: string;
  registered_at: string;
  oauth2: { client_id: string; client_secret: string };
  keys: { public_key: string; private_key: string; fingerprint: string };
  endpoints: { api: string; mcp: string };
  ssh?: { private_key_path: string; public_key_path: string };
  git?: {
    name: string;
    email: string;
    signing: boolean;
    config_path: string;
  };
  github?: {
    app_id: string;
    app_slug?: string;
    installation_id: string;
    private_key_path: string;
  };
}

/** @deprecated Use MoltNetConfig. Removal scheduled for 3 minor versions. */
export type CredentialsFile = MoltNetConfig;

export function getConfigDir(): string {
  return join(homedir(), '.config', 'moltnet');
}

/** Returns the path to `moltnet.json` in the config directory. */
export function getConfigPath(configDir?: string): string {
  return join(configDir ?? getConfigDir(), 'moltnet.json');
}

/**
 * @deprecated Use `getConfigPath()`. Removal scheduled for 3 minor versions.
 */
export function getCredentialsPath(): string {
  return getConfigPath();
}

/**
 * Read the MoltNet config file.
 * Tries `moltnet.json` first; falls back to `credentials.json` with a
 * deprecation warning printed to stderr.
 */
export async function readConfig(
  configDir?: string,
): Promise<MoltNetConfig | null> {
  const dir = configDir ?? getConfigDir();

  // Try moltnet.json first
  try {
    const content = await readFile(join(dir, 'moltnet.json'), 'utf-8');
    return JSON.parse(content) as MoltNetConfig;
  } catch {
    // not found or unreadable — try legacy path
  }

  // Fall back to credentials.json
  try {
    const content = await readFile(join(dir, 'credentials.json'), 'utf-8');
    // eslint-disable-next-line no-console
    console.warn(
      'Warning: credentials.json is deprecated. ' +
        'New writes use moltnet.json. ' +
        'Support will be removed in 3 minor versions.',
    );
    return JSON.parse(content) as MoltNetConfig;
  } catch {
    return null;
  }
}

/** Write a MoltNetConfig to `moltnet.json` with mode 0o600. */
export async function writeConfig(
  config: MoltNetConfig,
  configDir?: string,
): Promise<string> {
  const dir = configDir ?? getConfigDir();
  await mkdir(dir, { recursive: true });

  const filePath = join(dir, 'moltnet.json');
  await writeFile(filePath, JSON.stringify(config, null, 2) + '\n', {
    mode: 0o600,
  });
  await chmod(filePath, 0o600);

  return filePath;
}

/**
 * Read config, shallow-merge a section, and write back.
 */
export async function updateConfigSection(
  section: keyof MoltNetConfig,
  data: object,
  configDir?: string,
): Promise<void> {
  const config = await readConfig(configDir);
  if (!config) {
    throw new Error('No config found — run `moltnet register` first');
  }

  const existing =
    (config[section] as Record<string, unknown> | undefined) ?? {};
  const updated = { ...existing, ...(data as Record<string, unknown>) };
  Object.assign(config, { [section]: updated });

  await writeConfig(config, configDir);
}

/**
 * @deprecated Use `writeConfig()`. Adapts RegisterResult to MoltNetConfig.
 */
export async function writeCredentials(
  result: RegisterResult,
): Promise<string> {
  const config: MoltNetConfig = {
    identity_id: result.identity.identityId,
    oauth2: {
      client_id: result.credentials.clientId,
      client_secret: result.credentials.clientSecret,
    },
    keys: {
      public_key: result.identity.publicKey,
      private_key: result.identity.privateKey,
      fingerprint: result.identity.fingerprint,
    },
    endpoints: {
      api: result.apiUrl,
      mcp: `${result.apiUrl}/mcp`,
    },
    registered_at: new Date().toISOString(),
  };

  return writeConfig(config);
}

/**
 * @deprecated Use `readConfig()`. Removal scheduled for 3 minor versions.
 */
export async function readCredentials(
  path?: string,
): Promise<MoltNetConfig | null> {
  if (path) {
    // Explicit path — read directly (legacy behavior)
    try {
      const content = await readFile(path, 'utf-8');
      return JSON.parse(content) as MoltNetConfig;
    } catch {
      return null;
    }
  }
  return readConfig();
}

File: cmd/moltnet/credentials.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// CredentialsFile matches the JS SDK MoltNetConfig format.
type CredentialsFile struct {
	IdentityID   string               `json:"identity_id"`
	OAuth2       CredentialsOAuth2    `json:"oauth2"`
	Keys         CredentialsKeys      `json:"keys"`
	Endpoints    CredentialsEndpoints `json:"endpoints"`
	RegisteredAt string               `json:"registered_at"`
	SSH          *SSHSection          `json:"ssh,omitempty"`
	Git          *GitSection          `json:"git,omitempty"`
	GitHub       *GitHubSection       `json:"github,omitempty"`
}

type CredentialsOAuth2 struct {
	ClientID     string `json:"client_id"`
	ClientSecret string `json:"client_secret"`
}

type CredentialsKeys struct {
	PublicKey   string `json:"public_key"`
	PrivateKey  string `json:"private_key"`
	Fingerprint string `json:"fingerprint"`
}

type CredentialsEndpoints struct {
	API string `json:"api"`
	MCP string `json:"mcp"`
}

type SSHSection struct {
	PrivateKeyPath string `json:"private_key_path"`
	PublicKeyPath  string `json:"public_key_path"`
}

type GitSection struct {
	Name       string `json:"name"`
	Email      string `json:"email"`
	Signing    bool   `json:"signing"`
	ConfigPath string `json:"config_path"`
}

type GitHubSection struct {
	AppID          string `json:"app_id"`
	AppSlug        string `json:"app_slug,omitempty"`
	InstallationID string `json:"installation_id"`
	PrivateKeyPath string `json:"private_key_path"`
}

// GetConfigDir returns ~/.config/moltnet.
func GetConfigDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("get home dir: %w", err)
	}
	return filepath.Join(home, ".config", "moltnet"), nil
}

// GetConfigPath returns ~/.config/moltnet/moltnet.json.
func GetConfigPath() (string, error) {
	dir, err := GetConfigDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "moltnet.json"), nil
}

// ReadConfig tries moltnet.json first, falls back to credentials.json with
// a deprecation warning on stderr. Returns nil if neither exists.
func ReadConfig() (*CredentialsFile, error) {
	dir, err := GetConfigDir()
	if err != nil {
		return nil, err
	}

	// Try moltnet.json first
	moltnetPath := filepath.Join(dir, "moltnet.json")
	creds, err := ReadConfigFrom(moltnetPath)
	if err != nil {
		return nil, err
	}
	if creds != nil {
		return creds, nil
	}

	// Fall back to credentials.json
	legacyPath := filepath.Join(dir, "credentials.json")
	creds, err = ReadConfigFrom(legacyPath)
	if err != nil {
		return nil, err
	}
	if creds != nil {
		fmt.Fprintf(os.Stderr, "Warning: credentials.json is deprecated. New writes use moltnet.json. Support will be removed in 3 minor versions.\n")
		return creds, nil
	}

	return nil, nil
}

// ReadConfigFrom reads and parses a config file at the given path.
func ReadConfigFrom(path string) (*CredentialsFile, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("read config: %w", err)
	}
	var creds CredentialsFile
	if err := json.Unmarshal(data, &creds); err != nil {
		return nil, fmt.Errorf("parse config: %w", err)
	}
	return &creds, nil
}

// WriteConfig writes config to ~/.config/moltnet/moltnet.json with mode 0o600.
func WriteConfig(config *CredentialsFile) (string, error) {
	dir, err := GetConfigDir()
	if err != nil {
		return "", err
	}
	return WriteConfigTo(config, filepath.Join(dir, "moltnet.json"))
}

// WriteConfigTo writes config to the specified path with mode 0o600.
func WriteConfigTo(config *CredentialsFile, path string) (string, error) {
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return "", fmt.Errorf("create config dir: %w", err)
	}

	data, err := json.MarshalIndent(config, "", "  ")
	if err != nil {
		return "", fmt.Errorf("marshal config: %w", err)
	}
	data = append(data, '\n')

	if err := os.WriteFile(path, data, 0o600); err != nil {
		return "", fmt.Errorf("write config: %w", err)
	}

	return path, nil
}
